bnxt_re/lib: Support direct verbs

sbasavapatna opened the pull request:
Hi,

This patchset supports Direct Verbs in the bnxt_re library.

This is required by vendor specific applications that need to manage
the HW resources directly and to implement the datapath in the
application.

To support this, the library and the driver are being enhanced to
provide Direct Verbs using which the application can allocate and
manage the HW resources (Queues, Doorbell etc) . The Direct Verbs
enable the application to implement the control path.

Patch#1 Add a helper for VA-backed buffer descriptors
Patch#2 Export HSI file
Patch#3 Support DBR and UMEM verbs
Patch#4 Support CQ verbs
Patch#5 Support QP verbs
Patch#6 Add man pages
Patch#7 Add QP man pages

Thanks,
-Harsha

kaleshap86 committed:
bnxt_re/lib: Direct Verbs: Support DBR and UMEM verbs

The following Direct Verb (DV) APIs have been implemented in
this patch.

Doorbell Region Direct Verbs:

- bnxt_re_dv_alloc_db_region():
  This will allow the appliation to create extra doorbell regions
  and use the associated doorbell page index in dv_create_qp() and
  use the associated DB address while ringing the doorbell.

- bnxt_re_dv_free_db_region():
  Free the allocated doorbell region.

- bnxt_re_dv_get_default_db_region():
  Return the default doorbell page index and doorbell page address
  associated with the ucontext.

Umem Registration Direct Verbs:

- bnxt_re_dv_umem_reg():
  Register the user memory to be used by the application with
  the library. Application can register a large chunk of memory and
  use it during subsequent resource creation DV APIs.

  Note that the API terminates in the library and the app specified
  memory params (addr, len) are saved and a umem-handle is returned.
  That is, there is no ioctl to the driver at this point to map/pin
  the user memory.

  This memory is mapped/pinned later when the application creates
  the required resources (CQ/QP) using respective direct verbs.
  This is implemented in the next patch in this series.

- bnxt_re_dv_umem_dereg():
  Deregister the user memory specified by the umem-handle.

Co-developed-by: Sriharsha Basavapatna <sriharsha.basavapatna@broadcom.com>
Signed-off-by: Sriharsha Basavapatna <sriharsha.basavapatna@broadcom.com>
Signed-off-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Reviewed-by: Selvin Xavier <selvin.xavier@broadcom.com>

sbasavapatna committed:
bnxt_re/lib: Direct Verbs: Support CQ verbs

The following Direct Verb (DV) APIs have been implemented in
this patch. These are implemented by enhancing the driver
specific udata in existing verbs.

CQ Direct Verbs:

- bnxt_re_dv_create_cq():
  Create a CQ of requested size (cqe). The application must have
  already registered this memory using bnxt_re_dv_umem_reg().
  The CQ umem-handle and umem-offset provided by the application
  are translated into an address for mapping and passed to the
  driver.

- bnxt_re_dv_destroy_cq():
  Destroy the DV_CQ created earlier.

Note:
Some applications might want to allocate memory for all CQs in one big
chunk and then register that entire memory once using dv_umem_reg().
At the time of creating each individual resource, the application
should pass a specific offset/length in the umem registered memory.

Signed-off-by: Sriharsha Basavapatna <sriharsha.basavapatna@broadcom.com>
Co-developed-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Signed-off-by: Kalesh AP <kalesh-anakkur.purayil@broadcom.com>
Reviewed-by: Selvin Xavier <selvin.xavier@broadcom.com>

fix

jgunthorpe commented on providers/bnxt_re/bnxt_re_dv.h:
```diff
@@ -0,0 +1,78 @@
+			      struct bnxt_re_dv_db_region_attr *attr);
+int bnxt_re_dv_get_default_db_region(struct ibv_context *ibvctx,
+				     struct bnxt_re_dv_db_region_attr *out);
+void *bnxt_re_dv_umem_reg(struct ibv_context *ibvctx,
```
void *? That shouldn't be. Why do you have these functions when there is no kernel side?

jgunthorpe commented:
So give it an opaque type, don't just randomly use void *s. 

Again,why is this stuff here when there is no kernel side for umems?

sbasavapatna commented:
On Thu, Jan 8, 2026 at 1:28 AM Jason Gunthorpe ***@***.***>
wrote:

> *jgunthorpe* left a comment (linux-rdma/rdma-core#1654)
> <https://github.com/linux-rdma/rdma-core/pull/1654#issuecomment-3720537042>
>
> So give it an opaque type, don't just randomly use void *s.
>
Are you suggesting to use a pointer to the forward declaration of that
structure? Is that what you meant by "give it an opaque type"? Please
clarify.

> Again,why is this stuff here when there is no kernel side for umems?
>
The application registers the addr and the total size of the memory area
with the library, initially using this API. Later during each object
creation, the library uses this information to provide the actual va
(base-addr + offset) for mapping/pinning, to the kernel driver. This has
been captured in the kernel driver review thread:
***@***.***/

> —
> Reply to this email directly, view it on GitHub
> <https://github.com/linux-rdma/rdma-core/pull/1654#issuecomment-3720537042>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGTY2BIR6EFHFZXC2CENTI34FVQP5AVCNFSM6AAAAACK7AEXQ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTOMRQGUZTOMBUGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

sbasavapatna commented:
While looking into changes needed in the bnxt_re kernel driver to process DBR passed through driver specific attributes (during QP creation), I noticed that the corresponding libibverbs support seems to be missing in cmd_qp.c.
That is, ibv_cmd_create_qp_ex() doesn't support "ibv_command_buffer *driver" arg that should be linked to the main command buffer that it creates. This is supported in CQ creation command though (ibv_cmd_create_cq_ex2()).

sbasavapatna commented:
On Thu, Jan 15, 2026 at 10:19 PM Jason Gunthorpe ***@***.***>
wrote:

> *jgunthorpe* left a comment (linux-rdma/rdma-core#1654)
> <https://github.com/linux-rdma/rdma-core/pull/1654#issuecomment-3755833542>
>
> As the first user you'd need to do a similar thing as CQ.
>
Pushed the next version with a new create_qp function in libibverbs.
Thanks,
-Harsha

> —
> Reply to this email directly, view it on GitHub
> <https://github.com/linux-rdma/rdma-core/pull/1654#issuecomment-3755833542>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGTY2BPOYJBYLUJ4JB5E2LL4G7AIXAVCNFSM6AAAAACK7AEXQ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTONJVHAZTGNJUGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

sbasavapatna committed:
bnxt_re/lib: Direct Verbs: Support QP verbs

The following Direct Verb (DV) APIs have been implemented in
this patch. These are implemented by enhancing the driver
specific udata in existing verbs.

QP Direct Verbs:
- bnxt_re_dv_create_qp():
  Create a QP using specified params (struct bnxt_re_dv_qp_init_attr).
  The application must have already registered SQ/RQ memory using
  bnxt_re_dv_umem_reg(). The SQ/RQ umem-handle and umem-offset provided
  by the application are translated into an address for mapping and
  passed to the driver.

- bnxt_re_dv_destroy_qp():
  Destroy the DV_QP created earlier.

- bnxt_re_dv_modify_qp():
  Modify QP attributes of the DV_QP.

- bnxt_re_dv_query_qp():
  Return QP attributes of the DV_QP.

Note:
Some applications might want to allocate memory for all QPs in one big
chunk and then register that entire memory once using dv_umem_reg().
At the time of creating each individual resource, the application
should pass a specific offset/length in the umem registered memory.

Signed-off-by: Sriharsha Basavapatna <sriharsha.basavapatna@broadcom.com>
Reviewed-by: Selvin Xavier <selvin.xavier@broadcom.com>

sbasavapatna committed:
bnxt_re/lib: Direct Verbs: Add man pages

Document direct verbs (DV) feature and related DV APIs.
The man page describes these DV APIs:
- Doorbell regions (db)
- User memory (umem)
- Completion Queues (cq)

Signed-off-by: Sriharsha Basavapatna <sriharsha.basavapatna@broadcom.com>
Reviewed-by: Selvin Xavier <selvin.xavier@broadcom.com>

sbasavapatna committed:
bnxt_re/lib: Direct Verbs: Export HSI file

Extract HSI definitions (WQE/CQE structures and enums) from
bnxt_re-abi.h into a standalone bnxt_re-hsi.h and add it to
publish_headers() so it is installed under infiniband/.

This will make HSI definitions available to DV applications,
that implement their own datapath.

Signed-off-by: Sriharsha Basavapatna <sriharsha.basavapatna@broadcom.com>
Reviewed-by: Selvin Xavier <selvin.xavier@broadcom.com>

sbasavapatna committed:
bnxt_re/lib: Direct Verbs: Add QP man pages

Document QP related DV APIs:
    bnxt_re_dv_create_qp
    bnxt_re_dv_destroy_qp
    bnxt_re_dv_modify_qp
    bnxt_re_dv_query_qp

Signed-off-by: Sriharsha Basavapatna <sriharsha.basavapatna@broadcom.com>
Reviewed-by: Selvin Xavier <selvin.xavier@broadcom.com>

sbasavapatna committed:
verbs: Extend fill_attr_in_buf_umem() for VA-backed descriptors

fill_attr_in_buf_umem() only handles dmabuf-backed buffers.
Extend it for user virtual address buffers.

Signed-off-by: Sriharsha Basavapatna <sriharsha.basavapatna@broadcom.com>

jpirko commented on libibverbs/driver.h:
```diff
@@ -853,4 +853,25 @@ fill_attr_in_buf_umem(struct ibv_command_buffer *cmdb, uint16_t attr_id,
+	fill_attr_in_ptr(cmdb, attr_id, storage);
+	cmdb->fallback_ioctl_only = 1;
+}
+
```
Does not make sense to have this as separate helper. Your caller checks "buf->fd ==-1". You can do the check inside fill_attr_in_buf_umem and fill up things (like type) appropriately.

sbasavapatna replied:
Done, will update it in the next version.

jpirko replied:
looks okay to me now.

jpirko commented on providers/bnxt_re/bnxt_re_dv.h:
```diff
@@ -0,0 +1,78 @@
+};
+
+struct bnxt_re_dv_umem_reg_attr {
+	void *addr;
```
This duplicates the layout of struct ibv_buf. Why you can't use it directly?

sbasavapatna replied:
This is a user facing structure and we don't want to change it as it would impact applications that are already using it.

jpirko replied:
This puzzles me. How an application can use this struct if you are only introducing it now?

sbasavapatna replied:
OOB solution which already exists is being upstreamed.

jpirko replied:
Wait, you say that you form the API of a feature currently being pushed upon some out of tree code and use it as argument the API can't be changed? I have troubles to understand that reasoning, sorry.

sbasavapatna replied:
The structure layouts are slightly different. We have added a comp_mask for vendor specific extensions.
 
struct bnxt_re_dv_umem_reg_attr {
        void *addr;
        size_t size;
        uint64_t comp_mask;
        int dmabuf_fd;
};
struct ibv_buf {
        void *addr;
        size_t size;
        struct ibv_pd *pd;
        int dmabuf_fd;
};

Also,  since this is a Direct Verb to the user application, we want to maintain the same ABI for both our inbox and out-of-tree libraries, so that the applications can work seamlessly on both installations. So unless you see a design level issue here, we would prefer to keep this as-is and avoid any impact to existing customers that are using out-of-tree packages.

jpirko replied:
Why you have comp_mask here is a mystery. Not needed (and if really, really need it, you can embed ibv_buf - so no argument). Basically they are same structures. Don't screw-up the API in a name of out-of-tree code. That is fundamentally wrong. Please fix.

jgunthorpe replied:
I agree with Jiri, we have ibv_buf now that is the user facing API surface for a dmabuf or a VA or whatever, you should add some API to allow the user to create an ibv_buf from other things and consistently use that in your here in DV. Get rid of the umem here, it completely duplicates what ibv_buf is trying to do.

You guys choose to do the OOT thing, the broken ABIs are all your problem to deal with.

sbasavapatna replied:
Will add a couple of APIs as suggested to associate/disassociate an ibv_buf with user allocated memory. This ibv_buf will then be passed to bnxt_re_dv_create_cq/qp APIs.

jgunthorpe commented on providers/bnxt_re/dv.c:
```diff
@@ -0,0 +1,196 @@
+
+	ret = execute_ioctl(ibvctx, cmd);
+	if (ret) {
+		fprintf(stderr, "%s: execute_ioctl() failed: %d\n", __func__, ret);
```
Do not print errors from libraries like this, remove them all

sbasavapatna replied:
will remove these error prints.

jgunthorpe commented on providers/bnxt_re/verbs.c:
```diff
@@ -375,10 +377,15 @@ struct ibv_cq *bnxt_re_create_cq(struct ibv_context *ibvctx, int ncqe,
+				 (struct ibv_create_cq_ex *)&cmd.ibv_cmd,
+				 sizeof(cmd),
+				 (struct ib_uverbs_ex_create_cq_resp *)&resp.ibv_resp,
+				 sizeof(resp), 0))
```
Why all the casts here? It shouldn't have casts if written properly

sbasavapatna replied:
will fix this.

jgunthorpe commented on providers/bnxt_re/verbs.c:
```diff
@@ -2107,7 +2117,7 @@ static struct ibv_qp *__bnxt_re_create_qp(struct ibv_context *ibvctx,
 
 	if (ibv_cmd_create_qp_ex(ibvctx, &qp->vqp, attr,
-				&req.ibv_cmd, sizeof(req), &resp.ibv_resp, sizeof(resp)))
+				 &req.ibv_cmd, sizeof(req), &resp.ibv_resp, sizeof(resp)))
```
stray whitespace?

sbasavapatna replied:
accept.

jgunthorpe commented on providers/bnxt_re/dv.c:
```diff
@@ -194,3 +194,120 @@ int bnxt_re_dv_umem_dereg(struct bnxt_re_dv_umem *umem)
+				    (struct verbs_cq *)&cq->ibvcq,
+				    (struct ibv_create_cq_ex *)&cmd.ibv_cmd, sizeof(cmd),
+				    (struct ib_uverbs_ex_create_cq_resp *)&resp->ibv_resp,
+				    sizeof(*resp), 0, driver_attrs);
```
Also why casts?

sbasavapatna replied:
will be fixed.

jgunthorpe commented on providers/bnxt_re/dv.c:
```diff
@@ -311,3 +311,305 @@ int bnxt_re_dv_destroy_cq(struct ibv_cq *ibvcq)
+	dst->ah_attr.grh.flow_label = src->ah_attr.grh.flow_label;
+}
+
+int bnxt_re_dv_query_qp(struct ibv_qp *ibvqp, struct ib_uverbs_qp_attr *qp_attr)
```
This type ib_uverbs_qp_attr should not be returned from a library interface, it is a kernel facing type

I don't even understand why this function exists in the first place

sbasavapatna replied:
Thanks for catching this; in the next version this will be restored to standard bnxt_re_query_qp() and the dv function will be removed.

jgunthorpe commented on providers/bnxt_re/dv.c:
```diff
@@ -311,3 +311,305 @@ int bnxt_re_dv_destroy_cq(struct ibv_cq *ibvcq)
+
+	rc = ibv_cmd_query_qp(ibvqp, &attr, qp_attr->qp_attr_mask, &init_attr,
+			      &cmd, sizeof(cmd));
+	if (!rc) {
```
success oriented flow

sbasavapatna replied:
This function itself won't be there in the next version.

jgunthorpe commented on providers/bnxt_re/man/bnxt_re_dv_create_cq.3.md:
```diff
@@ -0,0 +1,57 @@
+registration, aligned to the device page size; **ncqe** is the requested number
+of CQEs.
+
+This path requires **IB_UVERBS_CORE_SUPPORT_ROBUST_UDATA** (robust udata) on the
```
Doesn't seem like something the user needs to know. You could write 'this call requires linux kernel version >= 'xyz

sbasavapatna replied:
done.

jgunthorpe commented on providers/bnxt_re/man/bnxt_re_dv_destroy_qp.3.md:
````diff
@@ -0,0 +1,46 @@
+```c
+#include <infiniband/bnxt_re_dv.h>
+
+int bnxt_re_dv_destroy_qp(struct ibv_qp *ibvqp);
````
Why does this exist? What is wrong with the normal QP destruction?

sbasavapatna replied:
DV variants of QP functions - destroy, query and modify will be removed; they were mostly wrappers.  Similarly dv_destroy_cq will also be removed.

mplaneta-qc commented on providers/bnxt_re/man/bnxt_re_dv_alloc_db_region.3.md:
```diff
@@ -0,0 +1,64 @@
+struct bnxt_re_dv_db_region_attr {
+        uint32_t  handle;   /* kernel handle for this doorbell region */
+        uint32_t  dpi;      /* doorbell page index */
+        uint64_t  umdbr;    /* unmapped doorbell BAR offset */
```
What is the intended flow for using `umdbr`? Does unmapped mean not mapped into application virtual address space?

sbasavapatna replied:
Yes, this is the raw PCI-BAR/physical-address of the doorbell register page.

mplaneta-qc replied:
How user space can use this field?

sbasavapatna replied:
It is not meant for normal user-space applications. It is intended to support future extensions for peer devices like GPUs that may need to directly write to the Doorbell BAR.

mplaneta-qc replied:
Is it supposed to be used somehow with DMA buf then? https://www.kernel.org/doc/html/latest/driver-api/dma-buf.html

jgunthorpe replied:
What? You cannot expose a physical address to userspace. Send a kernel patch to take this away please.

Aswin5005 replied:
Can we pass a File Descriptor with help of dma-buf framework to user space instead of returning raw Physical address ?

sbasavapatna replied:
> What? You cannot expose a physical address to userspace. Send a kernel patch to take this away please.

Since the ABI structure is already in released kernels (7.1, 7.2), we can't remove the field right? I have a patch that changes it to a reserved field and avoids setting it.  If that's ok, I can send out that patch, let me know.
